Java integration tests fail after changing to @EnableOAuth2Sso

Running `mvn verify` yields the following errors:

```
[ERROR] Tests run: 2, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 1.809 s <<< FAILURE! - in com.okta.developer.its.HoldingsApiWithMockMvcIT
[ERROR] testPostHoldings(com.okta.developer.its.HoldingsApiWithMockMvcIT)  Time elapsed: 0.079 s  <<< FAILURE!
java.lang.AssertionError: Response status expected:<200> but was:<403>
	at com.okta.developer.its.HoldingsApiWithMockMvcIT.testPostHoldings(HoldingsApiWithMockMvcIT.java:114)

2018-06-11 21:47:55.930  INFO 12723 --- [      Thread-27] ConfigServletWebServerApplicationContext : Closing org.springframework.boot.web.servlet.context.AnnotationConfigServletWebServerApplicationContext@3f60ab61: startup date [Mon Jun 11 21:47:54 MDT 2018]; root of context hierarchy
2018-06-11 21:47:55.930  INFO 12723 --- [       Thread-6] ConfigServletWebServerApplicationContext : Closing org.springframework.boot.web.servlet.context.AnnotationConfigServletWebServerApplicationContext@8c18bde: startup date [Mon Jun 11 21:47:48 MDT 2018]; root of context hierarchy
[INFO]
[INFO] Results:
[INFO]
[ERROR] Failures:
[ERROR]   HoldingsApiIT.testGetHoldings:94 Http response failed with non-2xx status code: 302
body:

[ERROR]   HoldingsApiIT.testPostHoldings:123 Http response failed with non-2xx status code: 302
body:

[ERROR]   HoldingsApiWithMockMvcIT.testPostHoldings:114 Response status expected:<200> but was:<403>
[INFO]
[ERROR] Tests run: 4, Failures: 3, Errors: 0, Skipped: 0
```

--- a/holdings-api/pom.xml
+++ b/holdings-api/pom.xml
@@ -85,6 +85,12 @@
             <version>2.8.0</version>
             <scope>test</scope>
         </dependency>
+        <dependency>
+            <groupId>io.rest-assured</groupId>
+            <artifactId>rest-assured</artifactId>
+            <version>3.0.3</version>
+            <scope>test</scope>
+        </dependency>
     </dependencies>
 
     <build>

--- a/holdings-api/src/test/java/com/okta/developer/holdingsapi/UserControllerTest.java
+++ b/holdings-api/src/test/java/com/okta/developer/holdingsapi/UserControllerTest.java
@@ -26,7 +26,10 @@ import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.
 import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;
 
 @RunWith(SpringRunner.class)
-@SpringBootTest(classes = HoldingsApiApplication.class)
+@SpringBootTest(classes = HoldingsApiApplication.class,
+                properties = {
+                    "okta.client.token=FAKE_TEST_TOKEN",
+                    "okta.client.orgUrl=https://example.com/not-used"})
 public class UserControllerTest {
 
     private MockMvc restUserMockMvc;

--- a/holdings-api/src/test/java/com/okta/developer/its/HoldingsApiIT.java
+++ b/holdings-api/src/test/java/com/okta/developer/its/HoldingsApiIT.java
@@ -51,8 +51,8 @@ import static org.springframework.boot.test.context.SpringBootTest.WebEnvironmen
                     "okta.oauth2.discoveryDisabled=true",
                     "okta.client.orgUrl=http://localhost:${wiremock.server.port}",
                     "okta.oauth2.issuer=http://localhost:${wiremock.server.port}/oauth/issuer",
-                    "okta.oauth2.clientId=" + HoldingsApiIT.TEST_CLIENT_ID,
-                    "okta.oauth2.clientSecret=" + HoldingsApiIT.TEST_CLIENT_SECRET,
+                    "security.oauth2.client.clientId=" + HoldingsApiIT.TEST_CLIENT_ID,
+                    "security.oauth2.client.clientSecret=" + HoldingsApiIT.TEST_CLIENT_SECRET,
                     "security.oauth2.resource.userInfoUri=http://localhost:${wiremock.server.port}/oauth/userInfoUri",
                     "security.oauth2.client.accessTokenUri=http://localhost:${wiremock.server.port}/oauth/token",
                     "security.oauth2.client.userAuthorizationUri=http://localhost:${wiremock.server.port}/oauth/authorize"

--- a/holdings-api/src/test/java/com/okta/developer/its/HoldingsApiIT.java
+++ b/holdings-api/src/test/java/com/okta/developer/its/HoldingsApiIT.java
@@ -4,39 +4,41 @@ import com.github.tomakehurst.wiremock.WireMockServer;
 import com.github.tomakehurst.wiremock.client.WireMock;
 import com.okta.developer.holdingsapi.Holding;
 import com.okta.developer.holdingsapi.HoldingsApiApplication;
+import io.restassured.http.ContentType;
+import io.restassured.response.ExtractableResponse;
 import org.junit.After;
-import org.junit.Assert;
 import org.junit.Before;
 import org.junit.Test;
 import org.junit.runner.RunWith;
-import org.springframework.beans.factory.annotation.Autowired;
 import org.springframework.beans.factory.annotation.Value;
 import org.springframework.boot.test.context.SpringBootTest;
-import org.springframework.boot.test.web.client.TestRestTemplate;
+import org.springframework.boot.web.server.LocalServerPort;
 import org.springframework.context.ApplicationContextInitializer;
 import org.springframework.context.ConfigurableApplicationContext;
 import org.springframework.http.MediaType;
-import org.springframework.http.client.ClientHttpResponse;
 import org.springframework.test.context.ContextConfiguration;
 import org.springframework.test.context.junit4.SpringRunner;
 import org.springframework.test.context.support.TestPropertySourceUtils;
 import org.springframework.util.SocketUtils;
 import org.springframework.util.StreamUtils;
-import org.springframework.web.client.ResponseErrorHandler;
 
-import javax.annotation.PostConstruct;
 import java.io.IOException;
 import java.nio.charset.StandardCharsets;
 import java.util.Arrays;
-import java.util.Collections;
 import java.util.List;
+import java.util.regex.Pattern;
 
 import static com.github.tomakehurst.wiremock.client.WireMock.aResponse;
+import static com.github.tomakehurst.wiremock.client.WireMock.containing;
+import static com.github.tomakehurst.wiremock.client.WireMock.matching;
+import static com.github.tomakehurst.wiremock.client.WireMock.urlPathEqualTo;
 import static com.github.tomakehurst.wiremock.core.WireMockConfiguration.wireMockConfig;
 import static com.spotify.hamcrest.pojo.IsPojo.pojo;
+import static io.restassured.RestAssured.given;
 import static org.hamcrest.MatcherAssert.assertThat;
 import static org.hamcrest.Matchers.contains;
 import static org.hamcrest.Matchers.is;
+import static org.hamcrest.text.MatchesPattern.matchesPattern;
 import static org.springframework.boot.test.context.SpringBootTest.WebEnvironment.RANDOM_PORT;
 
 @RunWith(SpringRunner.class)
@@ -49,49 +51,80 @@ import static org.springframework.boot.test.context.SpringBootTest.WebEnvironmen
                     "okta.oauth2.discoveryDisabled=true",
                     "okta.client.orgUrl=http://localhost:${wiremock.server.port}",
                     "okta.oauth2.issuer=http://localhost:${wiremock.server.port}/oauth/issuer",
-                    "security.oauth2.resource.userInfoUri=http://localhost:${wiremock.server.port}/oauth/userInfoUri"
+                    "okta.oauth2.clientId=" + HoldingsApiIT.TEST_CLIENT_ID,
+                    "okta.oauth2.clientSecret=" + HoldingsApiIT.TEST_CLIENT_SECRET,
+                    "security.oauth2.resource.userInfoUri=http://localhost:${wiremock.server.port}/oauth/userInfoUri",
+                    "security.oauth2.client.accessTokenUri=http://localhost:${wiremock.server.port}/oauth/token",
+                    "security.oauth2.client.userAuthorizationUri=http://localhost:${wiremock.server.port}/oauth/authorize"
                 })
 public class HoldingsApiIT {
 
     private final static String TEST_ACCESS_TOKEN = "fake-access-token";
     private final static String TEST_USER_EMAIl = "joe.coder@example.com";
     private final static String TEST_USER_ID = "user-id-123";
+    final static String TEST_CLIENT_ID = "FAKE_CLIENT_ID";
+    final static String TEST_CLIENT_SECRET = "FAKE_CLIENT_SECRET";
 
     private WireMockServer wireMockServer;
 
     @Value("${wiremock.server.port}")
     private int mockServerPort;
 
-    @Autowired
-    private TestRestTemplate restTemplate;
+    @LocalServerPort
+    int applicationPort;
 
-    @PostConstruct
-    private void configureRestTemplate() {
+    private ExtractableResponse performLogin() {
+        String expectedRedirect = Pattern.quote(
+                "http://localhost:" + mockServerPort + "/oauth/authorize" +
+                "?client_id=" + TEST_CLIENT_ID +
+                "&redirect_uri=http://localhost:" + applicationPort +"/login" +
+                "&response_type=code" +
+                "&scope=profile%20email%20openid" +
+                "&state=")+".{6}";
 
-        restTemplate.getRestTemplate().setInterceptors(Collections.singletonList((request, body, execution) -> {
-            request.getHeaders().add("Authorization", "Bearer "+ TEST_ACCESS_TOKEN);
-            return execution.execute(request, body);
-        }));
+        ExtractableResponse response1 = given()
+            .redirects()
+                .follow(false)
+            .accept(ContentType.JSON)
+        .when()
+            .get("http://localhost:" + applicationPort + "/login")
+        .then()
+            .statusCode(302)
+            .header("Location", matchesPattern(expectedRedirect))
+        .extract();
 
-        restTemplate.getRestTemplate().setErrorHandler(new ResponseErrorHandler() {
-            @Override
-            public boolean hasError(ClientHttpResponse response) throws IOException {
-                return !response.getStatusCode().is2xxSuccessful();
-            }
+        String redirectUrl = response1.header("Location");
+        String state = redirectUrl.substring(redirectUrl.lastIndexOf('=')+1);
+        String code = "TEST_CODE";
+        String requestUrl = "http://localhost:" + applicationPort + "/login?code=" + code + "&state=" + state;
 
-            @Override
-            public void handleError(ClientHttpResponse response) throws IOException {
-                String body = StreamUtils.copyToString(response.getBody(), StandardCharsets.UTF_8);
-                Assert.fail("Http response failed with non-2xx status code: " + response.getStatusCode() +"\n" +
-                        "body:\n" + body);
-            }
-        });
+        return given()
+            .accept(ContentType.JSON)
+            .cookies(response1.cookies())
+            .redirects()
+                .follow(false)
+        .when()
+            .get(requestUrl)
+        .then()
+            .statusCode(302)
+        .extract();
     }
 
     @Test
     public void testGetHoldings() {
 
-        List<Holding> holdings = Arrays.asList(restTemplate.getForObject("/api/holdings", Holding[].class));
+        ExtractableResponse response = given()
+            .accept(ContentType.JSON)
+            .cookies(performLogin().cookies())
+            .redirects()
+                .follow(false)
+        .when()
+            .get("http://localhost:" + applicationPort+"/api/holdings")
+        .then()
+            .statusCode(200)
+            .extract();
+
+        List<Holding> holdings = Arrays.asList(response.body().as(Holding[].class));
 
         // use Spotify's hamcrest-pojo to validate the objects
         assertThat(holdings, contains(
@@ -120,7 +153,23 @@ public class HoldingsApiIT {
                     .setAmount("amount-2")
         };
 
-        List<Holding> outputHoldings = Arrays.asList(restTemplate.postForObject("/api/holdings", inputHoldings, Holding[].class));
+        // get a vaild csrf token
+        ExtractableResponse loginResponse = performLogin();
+
+        ExtractableResponse response2 = given().log().everything() // TODO remove log
+            .accept(ContentType.JSON)
+            .cookie("JSESSIONID", loginResponse.cookies().get("JSESSIONID"))
+            .header("X-XSRF-TOKEN", loginResponse.cookies().get("XSRF-TOKEN"))
+            .body(inputHoldings)
+            .redirects()
+                .follow(false)
+        .when()
+            .post("http://localhost:" + applicationPort + "/api/holdings")
+        .then().log().everything()  // TODO remove log
+            .statusCode(200)
+            .extract();
+
+        List<Holding> outputHoldings = Arrays.asList(response2.body().as(Holding[].class));
 
         // output is the same as the input
         assertThat(outputHoldings, contains(
@@ -160,12 +209,29 @@ public class HoldingsApiIT {
         String body = StreamUtils.copyToString(getClass().getResourceAsStream("/its/user.json"), StandardCharsets.UTF_8);
 
         // respond to GET for user
-        wireMockServer.stubFor(WireMock.get("/api/v1/users/" + TEST_USER_EMAIl)
+        wireMockServer.stubFor(WireMock.get("/api/v1/users/" + TEST_USER_ID)
                 .willReturn(aResponse().withBody(body)));
 
         // respond to PUT for user
         wireMockServer.stubFor(WireMock.put("/api/v1/users/" + TEST_USER_ID)
                 .willReturn(aResponse().withBody(body)));
+
+        // OAuth token
+        wireMockServer.stubFor(
+                WireMock.post(urlPathEqualTo("/oauth/token"))
+                            .withRequestBody(containing("grant_type=authorization_code"))
+                            .withRequestBody(containing("code=TEST_CODE&"))
+                            .withRequestBody(matching(".*" + Pattern.quote("redirect_uri=http%3A%2F%2Flocalhost%3A") + "\\d+" + Pattern.quote("%2Flogin") + ".*"))
+                            .willReturn(aResponse()
+                            .withHeader("Content-Type", "application/json;charset=UTF-8")
+                            .withBody(
+                                "{" +
+                                  "\"access_token\":\"" + TEST_ACCESS_TOKEN + "\",\n" +
+                                  "\"token_type\":\"Bearer\",\n" +
+                                  "\"expires_in\":3600,\n" +
+                                  "\"scope\":\"profile openid email\",\n" +
+                                  "\"id_token\":\"idTokenjwt\"\n" +
+                                "}")));
 
         // OAuth userInfoUri
         String userInfoBody = StreamUtils.copyToString(getClass().getResourceAsStream("/its/userInfo.json"), StandardCharsets.UTF_8);
@@ -176,7 +242,6 @@ public class HoldingsApiIT {
                         .withBody(userInfoBody)
                         .withHeader("Content-Type", MediaType.APPLICATION_JSON_VALUE)
                 ));
-
     }
 
     public static class RandomPortInitializer implements ApplicationContextInitializer<ConfigurableApplicationContext> {

--- a/holdings-api/src/test/java/com/okta/developer/its/HoldingsApiIT.java
+++ b/holdings-api/src/test/java/com/okta/developer/its/HoldingsApiIT.java
@@ -60,7 +60,6 @@ import static org.springframework.boot.test.context.SpringBootTest.WebEnvironmen
 public class HoldingsApiIT {
 
     private final static String TEST_ACCESS_TOKEN = "fake-access-token";
-    private final static String TEST_USER_EMAIl = "joe.coder@example.com";
     private final static String TEST_USER_ID = "user-id-123";
     final static String TEST_CLIENT_ID = "FAKE_CLIENT_ID";
     final static String TEST_CLIENT_SECRET = "FAKE_CLIENT_SECRET";

--- a/holdings-api/src/test/java/com/okta/developer/its/HoldingsApiIT.java
+++ b/holdings-api/src/test/java/com/okta/developer/its/HoldingsApiIT.java
@@ -156,16 +156,17 @@ public class HoldingsApiIT {
         // get a vaild csrf token
         ExtractableResponse loginResponse = performLogin();
 
-        ExtractableResponse response2 = given().log().everything() // TODO remove log
+        ExtractableResponse response2 = given()
             .accept(ContentType.JSON)
-            .cookie("JSESSIONID", loginResponse.cookies().get("JSESSIONID"))
+            .contentType(ContentType.JSON)
+            .cookies(loginResponse.cookies())
             .header("X-XSRF-TOKEN", loginResponse.cookies().get("XSRF-TOKEN"))
             .body(inputHoldings)
             .redirects()
                 .follow(false)
         .when()
             .post("http://localhost:" + applicationPort + "/api/holdings")
-        .then().log().everything()  // TODO remove log
+        .then()
             .statusCode(200)
             .extract();
 

--- a/holdings-api/src/test/java/com/okta/developer/its/HoldingsApiWithMockMvcIT.java
+++ b/holdings-api/src/test/java/com/okta/developer/its/HoldingsApiWithMockMvcIT.java
@@ -37,6 +37,7 @@ import static com.github.tomakehurst.wiremock.client.WireMock.aResponse;
 import static com.github.tomakehurst.wiremock.core.WireMockConfiguration.wireMockConfig;
 
 import static org.springframework.boot.test.context.SpringBootTest.WebEnvironment.RANDOM_PORT;
+import static org.springframework.security.test.web.servlet.request.SecurityMockMvcRequestPostProcessors.csrf;
 import static org.springframework.security.test.web.servlet.request.SecurityMockMvcRequestPostProcessors.securityContext;
 import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.jsonPath;
 import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;
@@ -109,6 +110,7 @@ public class HoldingsApiWithMockMvcIT {
                 .contentType(MediaType.APPLICATION_JSON)
                 .accept(MediaType.APPLICATION_JSON)
                 .with(securityContext(SecurityContextHolder.getContext()))
+                .with(csrf())
                 .content(new ObjectMapper().writeValueAsString(holdings)))
 
                 .andExpect(status().is(200))

